### ISSUE TITLE
Libsigsegv-2.11 tests fail / cannot compile nix default environment

### ISSUE THREAD

**Issue by gdonval** [NixOS/nixpkgs#28464]
## Issue description

I am currently evaluating what Nix could bring to our high performance computing environment. It is an ancient environment without root access (see Technical details). There is no known problem between Libsigsegv-2.11 and the kernel I am forced to use (2.6.32 with latest security patches). This is a major problem as libsigsegv-2.11 is required to bootstrap nix.

Libsigsegv-2.11 tests fail with:

```raw
PASS: sigsegv1
checking for bison... bison -y
checking for bison... bison
checking for gm4... PASS: sigsegv2
no
checking for gnum4... no
checking for m4... m4
checking for flex... no
checking for lex... no
no
checking for flex... no
checking for makeinfo... checking for getloadavg... no
PASS: sigsegv3
/tmp/nix-build-binutils-2.28.drv-0/binutils-2.28/missing: line 81: makeinfo: command not found
checking for expect... no
checking for runtest... no
checking for ar... (cached) ar
checking for x86_64-unknown-linux-gnu-ar... (cached) ar
checking for as... (cached) as
checking for x86_64-unknown-linux-gnu-as... (cached) as
checking for x86_64-unknown-linux-gnu-dlltool... no
checking for dlltool... ../build-aux/test-driver: line 107: 38162 Bus error               "$@" > $log_file 2>&1
no
FAIL: stackoverflow1
checking for ld... (cached) ld
checking for x86_64-unknown-linux-gnu-ld... (cached) ld
yes
checking for x86_64-unknown-linux-gnu-lipo... yes
checking if gcc accepts -Wextra... no
checking for lipo... no
checking for nm... (cached) nm
checking for x86_64-unknown-linux-gnu-nm... (cached) nm
checking for ranlib... (cached) ranlib
checking for x86_64-unknown-linux-gnu-ranlib... (cached) ranlib
checking for strip... (cached) strip
checking for x86_64-unknown-linux-gnu-strip... (cached) strip
checking for x86_64-unknown-linux-gnu-windres... no
checking for windres... checking for string.h... no
checking for x86_64-unknown-linux-gnu-windmc... ../build-aux/test-driver: line 107: 38191 Bus error               "$@" > $log_file 2>&1
FAIL: stackoverflow2
no
checking for windmc... yes
checking sys/ioctl.h presence... no
checking for objcopy... (cached) objcopy
checking for x86_64-unknown-linux-gnu-objcopy... (cached) objcopy
checking for x86_64-unknown-linux-gnu-objdump... no
checking for objdump... objdump
checking for x86_64-unknown-linux-gnu-readelf... no
checking for readelf... readelf
checking for cc... cc
checking for c++... c++
checking for gcc... gcc
checking for gcj... no
checking for gfortran... no
checking for gccgo... yes
no
checking for features.h... yes
checking for sys/ioctl.h... yes
yes
checking if gcc accepts -Wvla... ============================
   : tests/test-suite.log
============================

# TOTAL: 5
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: stackoverflow1
====================

FAIL stackoverflow1 (exit status: 135)

FAIL: stackoverflow2
====================

FAIL stackoverflow2 (exit status: 135)

============================================================================
Testsuite summary for 
============================================================================
# TOTAL: 5
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0
============================================================================
See tests/test-suite.log
============================================================================
Makefile:727: recipe for target 'test-suite.log' failed
make[3]: *** [test-suite.log] Error 1
make[3]: Leaving directory '/tmp/nix-build-libsigsegv-2.11.drv-0/libsigsegv-2.11/tests'
Makefile:833: recipe for target 'check-TESTS' failed
make[2]: *** [check-TESTS] Error 2
make[2]: Leaving directory '/tmp/nix-build-libsigsegv-2.11.drv-0/libsigsegv-2.11/tests'
Makefile:962: recipe for target 'check-am' failed
make[1]: *** [check-am] Error 2
make[1]: Leaving directory '/tmp/nix-build-libsigsegv-2.11.drv-0/libsigsegv-2.11/tests'
Makefile:426: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1
```

### Steps to reproduce

* Install `nix` from source with a custom prefix: `/home/user/nix`.
* Setup repos: 
  ```
  $ source .../nix.sh
  $ nix-channel --remove nixpkgs 
  $ nix-channel --add https://nixos.org/channels/nixos-unstable
  $ nix-channel --update
  ```
* Install `bash`: 
  ```
  $ nix-env -i bash
  ```
* Bug...

## Technical details

* System:
  ```raw
  LSB Version:	:base-4.0-amd64:base-4.0-noarch:core-4.0-amd64:core-4.0-noarch:graphics-4.0-
  amd64:graphics-4.0-noarch:printing-4.0-amd64:printing-4.0-noarch
  Distributor ID:	Scientific
  Description:	Scientific Linux release 6.9 (Carbon)
  Release:	6.9
  Codename:	Carbon
  ```
* Nix version: nix-env (Nix) 1.11.13
* Nixpkgs version: nixos-unstable
* Sandboxing enabled: no

**Comment by vcunat** [comment]
I think you want to consider [using PRoot](https://web.archive.org/web/20160829175309/https://nixos.org/wiki/How_to_install_nix_in_home_(on_another_distribution)#PRoot_Installation).  One advantage is you could use our binaries, so as a side-effect you wouldn't have to build libsigsegv at all (unless you have some specific needs).

Another thing is that I suspect the disabled sandboxing might be triggering the problem, but I don't know if it's possible to enable it when using PRoot.

**Comment by dezgeg** [comment]
Are you sure your kernel isn't affected by this bug: https://github.com/NixOS/nixpkgs/issues/6028? The two failed tests are the same...

**Comment by vcunat** [comment]
Right, "*exit status: 135*" seemed likely a SIGBUS.

**Comment by gdonval** [comment]
Well, no I am not absolutely totally sure that my kernel isn't affected by that bug. However please keep in mind that:

* libsigsegv has only 5 tests... so the same 2 out of 5 could be a coincidence;
* those tests are the most brittle of all and the way it breaks is always the same even though the origin can be different;
* that specific problem started to occur with kernel `3.14.29` and was identified as a result of a nix update: it is unlikely that the problem occurred in previous versions;
* my system is ancient but security fixes are released regularly: the current version of libsigsegv on the system **is** 2.11 and as far as I can tell, it required no patch.

It seems very likely that sandboxing would solve the problem: the vendor could compile it on a box with a regular Linux disk layout. It is probable that there are a few hardcoded paths in-there. If I could make a patch to remove them, that would likely solve my problem and make nix bootstrapping process more robust.

**Comment by gdonval** [comment]
So I checked... with the supplied system binary library too...  That bug is indeed the same...

The patch cannot be found anymore and I cannot patch the kernel anyway... That's so irritating.

**Comment by gdonval** [comment]
So I could fix it... My C is not good but that case was easy enough.

I just had to change `src/signals.h`:

```
21c21
<   { int var; var = SIGSEGV; { body } }
---
>   { int var; var = SIGSEGV; { body } var = SIGBUS; { body } }
```

But this is where I need you: how can I apply that patch conditionally with host kernel number in `Nix`?

Also please keep in mind that this is occurring while **bootstrapping** `stdenv`: this must be an automated default.

**Comment by vcunat** [comment]
Host's kernel isn't accounted in derivation hashes, so acting based on that would be an impurity, and you might get surprises when using substitutes, copying to other hosts, changing kernels, etc.  I would most likely prefer some *explicit* option when desiring this workaround, perhaps `packageOverrides` or something like that.

**Comment by gdonval** [comment]
Of course, that's fine by me ! :) We could just put a sentence in the documentation afterwards to let people know about this...

If I suggested such an impure thing to begin with is because I was under the impression that `packageOverrides` wouldn't work. Actually, I thought at first that this bug was caused by an interaction between the older `gcc` version used by default by `stdenv`. So I tried to override the default `gcc5` with `gcc7` using `packageOverrides`. I understand more the step-by-step bootstrap process of `stdenv` now so I see why I was hitting a recursion infinite loop...

I also tried to use `libsigsegv-2.5` instead of `libsigsegv-2.11` to no avail, certainly because again that version is hardcoded in the bootstrap process.

So in our case, would a conditional application of the patch dependent on a new `fixOlderKernel ? false` flag be enough?

**Comment by vcunat** [comment]
I didn't realize it's involved in stdenv bootstrap; the standard overriding might not work there, I'm not sure.

Actually, ATM the libsigsegv usage in the bootstrap seems useless, as it's only used in gawk which for some reason loses the reference and therefore would most likely be unable to use the library.  After 8137a8cb738, neither stdenv nor nix will depend on libsigsegv (not even transitively).

**Comment by gdonval** [comment]
I have just committed a few changes to make that patch easy to enable. The standard overriding indeed won't work until that package is removed from `stdenv`.

**Comment by globin** [comment]
Should be closed by #28536, if not feel free to reopen
